### PR TITLE
Added a new method to validate the GitHub API URL

### PR DESCRIPTION
**Pull request by recena**
@reviewbybees specially, @oleg-nenashev

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
Hm... PR builder is dead

**Comment by recena** [comment]
@oleg-nenashev AFAIK, the test suite require local configuration. Was there a PR Builder configured in this repo?

**Comment by oleg-nenashev** [comment]
Yep. It has been configured on Kohsuke's beescloud account. I'm looking for it

File: pom.xml [hunk 1]
```diff
@@ -28,7 +28,7 @@
 
   <properties>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-    <findbugs-maven-plugin.version>3.0.1</findbugs-maven-plugin.version>
+    <findbugs-maven-plugin.version>3.0.2</findbugs-maven-plugin.version>
```
Comment (KostyaSha) [inline pom.xml]: Unrelated to commit feature change.

File: src/main/java/org/kohsuke/github/GitHub.java [hunk 2]
```diff
@@ -447,6 +456,23 @@ public boolean isCredentialValid() throws IOException {
     }
 
     /**
+     * Ensures that the API URL is valid.
+     */
+    public boolean isApiUrlValid() throws IOException {
+        try {
+            GHApiInfo apiInfo = retrieve().to("/", GHApiInfo.class);
+            try {
+                new URL(apiInfo.getRateLimitUrl());
```
Comment (KostyaSha) [inline src/main/java/org/kohsuke/github/GitHub.java]: Why api url is mixed to ratelimit?

File: src/main/java/org/kohsuke/github/GitHub.java [hunk 3]
```diff
@@ -205,6 +205,15 @@ public static GitHub connectAnonymously() throws IOException {
     }
 
     /**
+     * Connects to GitHub Enterprise anonymously.
```
Comment (KostyaSha) [inline src/main/java/org/kohsuke/github/GitHub.java]: What is the difference between anonymous and not? Why library should care about specific logins?

File: src/main/java/org/kohsuke/github/GitHub.java (hunk 2 above)
Comment (recena) [inline src/main/java/org/kohsuke/github/GitHub.java]: it is an easy way to verify that the endpoint (API URL) replied a right response. [Here](https://api.github.com/) an example.

File: src/main/java/org/kohsuke/github/GitHub.java (hunk 3 above)
Comment (recena) [inline src/main/java/org/kohsuke/github/GitHub.java]: One wants to connect to GH Enterprise as anonymous and this was not possible until now.

Comment (recena) [inline src/main/java/org/kohsuke/github/GitHub.java]: Analogous to `GitHub.connectAnonymously()`. [Reference](https://github.com/kohsuke/github-api/blob/master/src/main/java/org/kohsuke/github/GitHub.java#L203)

File: src/main/java/org/kohsuke/github/GitHub.java (hunk 2 above)
Comment (KostyaSha) [inline src/main/java/org/kohsuke/github/GitHub.java]: if rate_limit disappear this logic will be wrong.

File: pom.xml (hunk 1 above)
Comment (recena) [inline pom.xml]: Indeed, I'll wait to know the maintainer's opinion.

File: src/main/java/org/kohsuke/github/GitHub.java (hunk 3 above)
Comment (KostyaSha) [inline src/main/java/org/kohsuke/github/GitHub.java]: AFAIR Many of connect methods are useless and exist only for backward compatibility. Right logic should be use exist connects and builders. I see no reason for making separate methods or using anonymous in any env at all.

Comment (KostyaSha) [inline src/main/java/org/kohsuke/github/GitHub.java]: > Analogous to GitHub.connectAnonymously(). Reference

So you can't call `new GitHubBuilder().build()` from your code??

File: pom.xml (hunk 1 above)
Comment (KostyaSha) [inline pom.xml]: Never mix 2 unrelated changes in one when it possible. FB absolutely not related to `Added a new method to validate the GitHub API URL`

File: src/main/java/org/kohsuke/github/GitHub.java [hunk 4]
```diff
@@ -205,6 +205,15 @@ public static GitHub connectAnonymously() throws IOException {
...
+     *
+     * All operations that requires authentication will fail.
+     */
+    public static GitHub connectToEnterpriseAnonymously(String apiUrl) throws IOException {
```
Comment (lanwen) [inline src/main/java/org/kohsuke/github/GitHub.java]: -1 for new method. Builder exists for such cases

File: pom.xml (hunk 1 above)
Comment (oleg-nenashev) [inline pom.xml]: I agree with @KostyaSha

File: src/main/java/org/kohsuke/github/GitHub.java (hunk 2 above)
Comment (recena) [inline src/main/java/org/kohsuke/github/GitHub.java]: The same way than if disappear other attributes. This method will return `null` is this case. By the way, when it happens GitHub API v4 will be published and a review of this client will be required.

Comment (KostyaSha) [inline src/main/java/org/kohsuke/github/GitHub.java]: method called `isApiUrlValid()` if it fetches and deserialized -> then true. 
If rate_limit valid -> other logic that can be checked even now.

File: src/main/java/org/kohsuke/github/GitHub.java (hunk 3 above)
Comment (recena) [inline src/main/java/org/kohsuke/github/GitHub.java]: Well, in this case we use the endpoint of a GH Enterprise server. I don't know where is the problem to add a new method.

File: src/main/java/org/kohsuke/github/GitHub.java (hunk 2 above)
Comment (lanwen) [inline src/main/java/org/kohsuke/github/GitHub.java]: why not just check status code and headers like
```
X-GitHub-Media-Type: github.v3
X-GitHub-Request-Id: 64c9ffd2-29dc-4a81-a85c-57fdba8369a4
```

File: src/main/java/org/kohsuke/github/GitHub.java (hunk 4 above)
Comment (recena) [inline src/main/java/org/kohsuke/github/GitHub.java]: But, it needs to provide the `apiUrl`.

File: src/main/java/org/kohsuke/github/GitHub.java (hunk 3 above)
Comment (KostyaSha) [inline src/main/java/org/kohsuke/github/GitHub.java]: This methods existed before normal builder appeared, use builder from your code as valid way of connection creation.

File: src/main/java/org/kohsuke/github/GitHub.java (hunk 2 above)
Comment (recena) [inline src/main/java/org/kohsuke/github/GitHub.java]: Because these header are returned in any service of GitHub API. And the goal of this method is to verify if `apiUrl` is really a valid endpoint.

File: src/main/java/org/kohsuke/github/GitHub.java (hunk 3 above)
Comment (recena) [inline src/main/java/org/kohsuke/github/GitHub.java]: I'm sorry but I don't understand your last comment. An example?

File: pom.xml (hunk 1 above)
Comment (recena) [inline pom.xml]: @oleg-nenashev @KostyaSha I agree. I'll change it.

File: src/main/java/org/kohsuke/github/GitHub.java (hunk 4 above)
Comment (lanwen) [inline src/main/java/org/kohsuke/github/GitHub.java]: so why not to add a new method for every combination of builder methods? :)

Comment (recena) [inline src/main/java/org/kohsuke/github/GitHub.java]: It could be other option. I prefer this.

**Comment by kohsuke** [comment]
I have a mixed feeling about adding more `connectXyz` methods, but it is indeed convenient to have some of them. Given that @recena feels strongly about having this particular convenience method, I think it's OK to have it.

I've massaged this PR somewhat by the following commit.

**Comment by recena** [comment]
@kohsuke Thanks for your time. I left `GHApiInfo` in a public class because we can need adding more fields in the future. I used `isApiUrlValid()` because we had `isCredentialsValid()`, I tried to follow the rules.
